### PR TITLE
generalize vec->store rule [run_process_replay] [no_assert]

### PR DESCRIPTION
**Pull request by DKormann**

### CODE REVIEW

**Comment by Qazalin** [comment]
Is there a fundamental reason this is limited to UOps.STORE?
Anything that goes from float4 -> gep -> float4 could fold the gep. eg: `METAL=1 python3 test/test_linearizer.py TestKernelOpts.test_tensor_core_opts`

```
    // float2 acc0 = float2(0.0f,0.0f)
    float2 wmma3 = __WMMA_8_8_8_float_float(float2(val3.x,val3.y), cast0, float2(acc0.x,acc0.y));
    acc0 = float2(wmma3.x,wmma3.y);
```

File: test/test_uop_graph.py [hunk 1]
```diff
@@ -128,6 +128,20 @@ def test_const_vectorize_fold(self):
     self.assertEqual(out.op, UOps.CONST)
     self.assertEqual(out.arg, 0.0)
 
+  def test_unvectorize_sizes(self):
+    uop = UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(
```
Comment (Qazalin) [inline test/test_uop_graph.py:132]: partial UOps are fine for UPats but tests should have a complete UOp.

File: test/test_uop_graph.py [hunk 2]
```diff
@@ -128,6 +128,20 @@ def test_const_vectorize_fold(self):
...
+      UOp(UOps.GEP, arg=i, src=(UOp.var("val"),)) for i in range(4))))
+
+    uop3 = UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(
+      UOp(UOps.GEP, arg=i, src=(UOp.var("val"),)) for i in range(3))))
+
+    uop_ = UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(
+      UOp(UOps.GEP, arg=0, src=(UOp.var("val"),)) for _ in range(4))))
+
+    self.assertEqual(constant_folder.rewrite(uop).op, UOps.STORE)
```
Comment (Qazalin) [inline test/test_uop_graph.py:141]: can you test UOpGraph([]).linearize() end-to-end instead?

**Comment by DKormann** [comment]
generalized to vectorize(gep(val) * n  ) -> val

**Comment by DKormann** [comment]
generalized to vectorize(gep(val) * n) -> val

File: tinygrad/codegen/uops.py [hunk 3]
```diff
@@ -305,11 +304,9 @@ def loop_collapse(loop_start, loop_end, compval, idx, mval, multconst, rng):
   # TODO: can do the invert of this (flip alt/load) when we fix double ops
   (UOp.store(UOp.var("buf"), UOp.var("idx"), UOp.alu(TernaryOps.WHERE, UOp.var("gate"), UOp.var("alt"), UOp.load(UOp.var("buf"), UOp.var("idx")))),
    lambda buf, idx, gate, alt: UOp.store(buf, idx, alt, gate)),
-  # store float4/float2 directly (remove VECTORIZE/GEP)
-  (UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(
-    UOp(UOps.GEP, arg=i, src=(UOp.var("val"),)) for i in range(4)))), UOp.store),
-  (UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(
-    UOp(UOps.GEP, arg=i, src=(UOp.var("val"),)) for i in range(2)))), UOp.store),
+  # Anything that goes from float4 -> gep -> float4 can fold the gep
+  (UPat(UOps.VECTORIZE, name="root", src=UPat(UOps.GEP, src=UPat(name='val')), allow_any_len=True),
```
Comment (geohot) [inline tinygrad/codegen/uops.py:308]: Err, what if val is a different type (larger)?

**Comment by Qazalin** [comment]
completed in Lowerer.

**Pull request closed without merging**